cf-check/replicate_lmdb.c

```diff
@@ -0,0 +1,270 @@
+
+    rc = mdb_env_create(&s_env);
+    if (rc != 0)
+    {
+        ret = rc;
+        report_mdb_error(s_file, rc);
+        goto cleanup;
+    }
+    mdb_env_set_userctx(s_env, &info);
+    mdb_env_set_assert(s_env, (MDB_assert_func*) HandleSrcLMDBCorruption);
+
+    /* MDB_NOTLS -- Don't use Thread-Local Storage
+     * Should allow us to use multiple transactions from the same thread (only
+     * one read-write). */
+    rc = mdb_env_open(s_env, s_file, MDB_NOSUBDIR | MDB_RDONLY | MDB_NOTLS, 0600);
+    if (rc != 0)
+    {
+        ret = rc;
+        report_mdb_error(s_file, rc);
+        goto cleanup;
+    }
+
+    rc = mdb_txn_begin(s_env, NULL, MDB_RDONLY, &s_txn);
+    if (rc != 0)
+    {
+        ret = rc;
+        report_mdb_error(s_file, rc);
+        goto cleanup;
+    }
+    info.s_txn = s_txn;
+
+    rc = mdb_dbi_open(s_txn, NULL, 0, &s_dbi);
+    if (rc != 0)
+    {
+        ret = rc;
+        report_mdb_error(s_file, rc);
+        goto cleanup;
+    }
+    else
+    {
+        close_s_dbi = true;
+    }
+
+    rc = mdb_cursor_open(s_txn, s_dbi, &s_cursor);
+    if (rc != 0)
+    {
+        ret = rc;
+        report_mdb_error(s_file, rc);
+        goto cleanup;
+    }
+
+    rc = mdb_env_create(&d_env);
+    if (rc != 0)
+    {
+        ret = rc;
+        report_mdb_error(d_file, rc);
+        goto cleanup;
+    }
+    mdb_env_set_userctx(d_env, &info);
+    mdb_env_set_assert(d_env, (MDB_assert_func*) HandleDstLMDBCorruption);
+
+    rc = mdb_env_open(d_env, d_file, MDB_NOSUBDIR | MDB_NOTLS, 0600);
+    if (rc != 0)
+    {
+        ret = rc;
+        report_mdb_error(d_file, rc);
+        goto cleanup;
+    }
+
+    rc = mdb_txn_begin(d_env, NULL, 0, &d_txn);
+    if (rc != 0)
+    {
+        ret = rc;
+        report_mdb_error(d_file, rc);
+        goto cleanup;
+    }
+    info.d_txn = d_txn;
+
+    rc = mdb_dbi_open(d_txn, NULL, MDB_CREATE, &d_dbi);
+    if (rc != 0)
+    {
+        ret = rc;
+        report_mdb_error(d_file, rc);
+        goto cleanup;
+    }
+    else
+    {
+        close_d_dbi = true;
+    }
+
+    rc = mdb_cursor_open(d_txn, d_dbi, &d_cursor);
+    if (rc != 0)
+    {
+        ret = rc;
+        report_mdb_error(d_file, rc);
+        goto cleanup;
+    }
+
+    rc = MDB_SUCCESS;
+    while (rc == MDB_SUCCESS)
+    {
+        MDB_val key, data;
+
+        rc = mdb_cursor_get(s_cursor, &key, &data, MDB_NEXT);
+        /* MDB_NOTFOUND => no more data */
+        if ((rc != MDB_SUCCESS) && (rc != MDB_NOTFOUND))
+        {
+            report_mdb_error(s_file, rc);
+            ret = rc;
+        }
+        if (rc == MDB_SUCCESS)
+        {
+            rc = mdb_put(d_txn, d_dbi, &key, &data, 0);
+            if (rc != MDB_SUCCESS)
+            {
+                report_mdb_error(d_file, rc);
+                ret = rc;
+            }
+        }
+    }
+    mdb_txn_commit(d_txn);
+    d_txn = NULL;
+    info.d_txn = NULL;
+
+  cleanup:
+    if (s_cursor != NULL)
+    {
```

Comment:
Any need to pay attention to return values for all these cleanup function calls?

Reply:
There's nothing that can be done in such cases apart from just keeping going. So I don't think it would be worth complicating the code and making it even longer.